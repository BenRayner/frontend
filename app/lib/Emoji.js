const APPLE_EMOJIS = require("!!./../../webpack/emoji-loader!./../../vendor/emojis/img-apple-64.json");
const BUILDKITE_EMOJIS = require("!!./../../webpack/emoji-loader!./../../vendor/emojis/img-buildkite-64.json");

const UNICODE_REGEXP = new RegExp('\\ud83c[\\udf00-\\udfff]|\\ud83d[\\udc00-\\ude4f]|\\ud83d[\\ude80-\\udeff]', 'g');
const COLON_REGEXP = new RegExp('\:[^\\s:]+\:', 'g');

class Emoji {
  parse(string, options = {}) {
    if(!string || !string.length) {
      return "";
    }

    // Start with replacing BK emojis (which are more likely than Apple emojis)
    string = this._replace(BUILDKITE_EMOJIS, COLON_REGEXP, string);

    // Then do an Apple emoji parse
    string = this._replace(APPLE_EMOJIS, UNICODE_REGEXP, string);
    string = this._replace(APPLE_EMOJIS, COLON_REGEXP, string);

    return string;
  }

  _replace(catalogue, regexp, string) {
    let matches = string.match(regexp);
    let replacements = [];

    // Bail if there aren't any emojis to replace
    if(!matches || !matches.length) {
      return string;
    }

    for(let i = 0, l = matches.length; i < l; i++) {
      let match = matches[i];
      let nextMatch = matches[i+1];

      // See if this match and the next one, makes a new emoji. For example,
      // :fist::skin-tone-4:
      if(nextMatch) {
        let matchWithModifier = `${match}${nextMatch}`
        let modifiedEmojiIndex = catalogue.indexed[matchWithModifier];

        if(modifiedEmojiIndex) {
          replacements.push(this._image(catalogue, modifiedEmojiIndex));
          replacements.push("");
          i += 1
        }
      }

      let emojiIndex = catalogue.indexed[match];

      if(emojiIndex) {
        replacements.push(this._image(catalogue, emojiIndex));
      } else {
        replacements.push(match);
      };
    };

    return string.replace(regexp, (match) => {
      return replacements.shift();
    });
  }

  _image(catalogue, index) {
    let emoji = catalogue.emojis[index];

    return `<img class="emoji" title="${emoji.name}" alt="${emoji.name}" src="${catalogue.host}/${emoji.image}" height="20" width="20" align="absmiddle" />`
  }
}

export default new Emoji();
