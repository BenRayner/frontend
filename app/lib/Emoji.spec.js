/* global describe, it, expect, jest */
import Emoji from './Emoji';

function mockLoadWebpackedEmojis(catalogue) {
  process.env.EMOJI_HOST = 'emoji-host.com';

  const loader = require('../../webpack/emoji-loader');
  const emojis = require(`../../vendor/emojis/img-${catalogue}-64.json`);
  const webpack = {};

  return eval(loader.call(webpack, emojis)); // eslint-disable-line no-eval
}

jest.mock('../emojis/buildkite', () => mockLoadWebpackedEmojis('buildkite'));
jest.mock('../emojis/apple', () => mockLoadWebpackedEmojis('apple'));

describe('Emoji', () => {
  describe('parse', () => {
    it('turns emojis into HTML', () => {
      expect(Emoji.parse(":buildkite:")).toMatchSnapshot();
      expect(Emoji.parse(":wave::skin-tone-3:")).toMatchSnapshot();
      expect(Emoji.parse("👋🏿")).toMatchSnapshot();
    });

    it('allows you to turn of unicode parsing of emojis', () => {
      expect(Emoji.parse("👍 :buildkite:", { replaceUnicode: false })).toMatchSnapshot();
    });

    it('escapes other HTML by default', () => {
      expect(Emoji.parse(":tada: <strong>This be strong...</strong>")).toMatchSnapshot();
    });

    it('allows you to turn off HTML escaping', () => {
      expect(Emoji.parse(":tada: <strong>This be strong...</strong>", { escape: false })).toMatchSnapshot();
    });
  });
});
