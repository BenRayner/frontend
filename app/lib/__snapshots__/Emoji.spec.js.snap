// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoji parse allows you to turn off HTML escaping 1`] = `"🎉 <strong>This be strong...</strong>"`;

exports[`Emoji parse converts emoji shortcodes ":buildkite:" 1`] = `"<img class=\\"emoji\\" title=\\"buildkite\\" alt=\\":buildkite:\\" src=\\"emoji-host.com/img-buildkite-64/buildkite.png\\" draggable=\\"false\\" />"`;

exports[`Emoji parse converts emoji shortcodes ":wave::skin-tone-3:" 1`] = `"👋🏼"`;

exports[`Emoji parse converts emoji shortcodes ":woman-woman-girl:" 1`] = `"👩‍👩‍👧"`;

exports[`Emoji parse converts emoji shortcodes "String with :rocket:" 1`] = `"String with 🚀"`;

exports[`Emoji parse escapes other HTML by default 1`] = `"🎉 &lt;strong&gt;This be strong...&lt;/strong&gt;"`;

exports[`Emoji parse ignores escaped shortcodes "String with \\:rocket\\:" 1`] = `"String with \\\\:rocket\\\\:"`;

exports[`Emoji parse ignores escaped shortcodes "arn\\:aws\\:s3:::bucket" 1`] = `"arn\\\\:aws\\\\:s3:::bucket"`;

exports[`Emoji parse leaves Unicode emoji alone "©️©" 1`] = `"©️©"`;

exports[`Emoji parse leaves Unicode emoji alone "™" 1`] = `"™"`;

exports[`Emoji parse leaves Unicode emoji alone "⛄" 1`] = `"⛄"`;

exports[`Emoji parse leaves Unicode emoji alone "🅱️epis" 1`] = `"🅱️epis"`;

exports[`Emoji parse leaves Unicode emoji alone "🇦🇺💜🇨🇦" 1`] = `"🇦🇺💜🇨🇦"`;

exports[`Emoji parse leaves Unicode emoji alone "👋🏿" 1`] = `"👋🏿"`;

exports[`Emoji parse leaves Unicode emoji alone "👍 :buildkite:" 1`] = `"👍 <img class=\\"emoji\\" title=\\"buildkite\\" alt=\\":buildkite:\\" src=\\"emoji-host.com/img-buildkite-64/buildkite.png\\" draggable=\\"false\\" />"`;

exports[`Emoji parse leaves Unicode emoji alone "👩‍👩‍👧" 1`] = `"👩‍👩‍👧"`;

exports[`Emoji parse leaves Unicode emoji alone "👩🏻‍🏫" 1`] = `"👩🏻‍🏫"`;

exports[`Emoji parse leaves Unicode emoji alone "💩 in the 💨" 1`] = `"💩 in the 💨"`;
